### PR TITLE
[libc++][TZDB] Implements time_zone get_info(local_time).

### PR DESCRIPTION
**Pull request by mordante**
Implements parts of:
- P0355 Extending chrono to Calendars and Time Zones

### CODE REVIEW

File: libcxx/src/time_zone.cpp [hunk 1]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+   *
+   * At the first of November the time goes backward one hour. This means the
+   * time marked with ^^ happens twice. This is represented by the ambiguous
+   * value in local_info.result.
+   *
+   * 2020.11.01                  2021.04.01              2021.11.01
+   * offset +05                  offset +05              offset +05
+   * save    0s                  save    1h              save    0s
+   * |-------------W----------|
+   *                             |----------W--------------|
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:997]:
```suggestion
   * |-------------//---------|
   *                             |----------//-------------|
```

That might make it clearer that this is meant to represent "a jump on the axis".

File: libcxx/src/time_zone.cpp [hunk 2]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+   *                                                    |-------------
+   *                           ~~                        ^^
+   *
+   * These shifts can happen due to changes in the current time zone for a
+   * location. For example, Indian/Kerguelen switched only once. In 1950 from an
+   * offset of 0 hours to an offset of +05 hours.
+   *
+   * During all these shifts the UTC time will have not gaps.
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1005]:
```suggestion
   * During all these shifts the UTC time will not have gaps.
```

File: libcxx/src/time_zone.cpp [hunk 3]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+   */
+
+  // The code needs to determine the system time for the local time. There is no
+  // information available. Assume the offset between system time and local time
+  // is 0s. This gives an initial estimate.
+  sys_seconds __guess{__local_seconds};
+  sys_info __info = __get_info(__guess);
+
+  // At this point the offset can be used to determine an estimate for the local
+  // time. Before doing the determine the offset validate whether the local time
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1015]:
```suggestion
  // time. Before doing that, determine the offset and validate whether the local time
```

File: libcxx/src/time_zone.cpp [hunk 4]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+  // is the range [chrono::local_seconds::min(), chrono::local_seconds::max()).
+  if (__local_seconds < 0s && __info.offset > 0s)
+    if (__local_seconds - chrono::local_seconds::min().time_since_epoch() < __info.offset)
+      return {-1, __info, {}};
+
+  if (__local_seconds > 0s && __info.offset < 0s)
+    if (chrono::local_seconds::max().time_since_epoch() - __local_seconds < -__info.offset)
+      return {-2, __info, {}};
+
+  // Based on the information in the sys_info found the local time can be
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1025]:
```suggestion
  // Based on the information found in the sys_info, the local time can be
```

File: libcxx/src/time_zone.cpp [hunk 5]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+  // converted to a system time. This resulting time can be in the following
+  // locations of the sys_info found:
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1027]:
```suggestion
  // locations of the sys_info:
```

File: libcxx/src/time_zone.cpp [hunk 6]
```diff
@@ -903,6 +904,180 @@ time_zone::__get_info(sys_seconds __time) const {
...
+  // 1. The estimate is before the returned sys_info object.
+  //    The result is either non-existent or unique in the previous sys_info.
+  // 2. The estimate is in the beginning of the returned sys_info object.
+  //    The result is either unique or ambiguous with the previous sys_info.
+  // 3. The estimate is in the "middle" of the returned sys_info.
+  //    The result is unique.
+  // 4. The result is at the end of the returned sys_info object.
+  //    The result is either unique or ambiguous with the next sys_info.
+  // 5. The estimate is after the returned sys_info object.
+  //    The result is either non-existent or unique in the next sys_info.
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1041]: The notion of "at the beginning of a sys info object" and "at the end of [...]" seems a bit arbitrary to me. Naively, I would instead say that we are either "within" a sys info object, or not. I would basically:

1. Is the time before the "current" sys_info object? If so, non-existent or unique in the previous one.
2. Is the time within the "current" sys_info object? If so, check whether that time is also in the previous one (in which case it's ambiguous), or if that time is also in the next one (ambiguous too). If not, that time is uniquely within the "current" sys_info.
3. Is the time after the "current" sys_info? If so, non-existent or unique in the next.

Doing it this way would be more expensive cause we have to materialize the previous/next sys_infos, but I think it's the correct algorithm to check what we want here. And optimizations to materializing sys_infos can be done to make this more efficient (such optimizations would be useful on their own anyway).

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/src/time_zone.cpp (hunk 6 above)
Comment (mordante) [reply libcxx/src/time_zone.cpp:1041]: I'm not too fond of your proposed wording, I've used different wording.
The implementation now no longer detects a "middle". It now tests the previous and next sys_info object.

File: libcxx/src/time_zone.cpp [hunk 7]
```diff
@@ -903,6 +904,152 @@ time_zone::__get_info(sys_seconds __time) const {
...
+  //
+  // 1. The estimate is before the returned sys_info object.
+  //    The result is either non-existent or unique in the previous sys_info.
+  // 2. The estimate is in the sys_info object
+  //    - If the sys_info begin is not sys_seconds::min(), then it might be at
+  //      2.1 and could be ambiguous with the previous or unique.
+  //    - If sys_info end is not sys_seconds::max(), then it might be at 2.3
+  //      and could be ambiguous with the next or unique.
+  //    - Else it is at 2.2 and always unique. This case happens when a
+  //      time zone has not transitions. For example, UTC or GMT+1.
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1011]:
```suggestion
  //      time zone has no transitions. For example, UTC or GMT+1.
```

File: libcxx/src/time_zone.cpp [hunk 8]
```diff
@@ -903,6 +904,152 @@ time_zone::__get_info(sys_seconds __time) const {
...
+  if (__local_seconds > 0s && __info.offset < 0s)
+    if (chrono::local_seconds::max().time_since_epoch() - __local_seconds < -__info.offset)
+      return {-2, __info, {}};
+
+  // Based on the information found in the sys_info, the local time can be
+  // converted to a system time. This resulting time can be in the following
+  // locations of the sys_info:
+  //
+  //                             |---------//--------------|
+  //                           1   2.1      2.2         2.3  5
```
Comment (ldionne) [inline libcxx/src/time_zone.cpp:1001]:
```suggestion
  //                           1   2.1      2.2         2.3  3
```

**Review by ldionne — APPROVED** [review]

**Comment by zeroomega** [comment]
We started to see flakiness in test `llvm-libc++-static.cfg.in :: std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp` in libcxx testing on linux-x64, after this change was landed

```
Exit Code: 250

Command Output (stdout):
--
# COMPILED WITH
/b/s/w/ir/x/w/llvm_build/./bin/clang++ /b/s/w/ir/x/w/llvm-llvm-project/libcxx/test/std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp -pthread --target=x86_64-unknown-linux-gnu -nostdinc++ -I /b/s/w/ir/x/w/llvm_build/include/c++/v1 -I /b/s/w/ir/x/w/llvm_build/include/x86_64-unknown-linux-gnu/c++/v1 -I /b/s/w/ir/x/w/llvm-llvm-project/libcxx/test/support -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings  -lc++experimental -nostdlib++ -L /b/s/w/ir/x/w/llvm_build/./lib/x86_64-unknown-linux-gnu -lc++ -lc++abi -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir/t.tmp.exe
# executed command: /b/s/w/ir/x/w/llvm_build/./bin/clang++ /b/s/w/ir/x/w/llvm-llvm-project/libcxx/test/std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp -pthread --target=x86_64-unknown-linux-gnu -nostdinc++ -I /b/s/w/ir/x/w/llvm_build/include/c++/v1 -I /b/s/w/ir/x/w/llvm_build/include/x86_64-unknown-linux-gnu/c++/v1 -I /b/s/w/ir/x/w/llvm-llvm-project/libcxx/test/support -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -lc++experimental -nostdlib++ -L /b/s/w/ir/x/w/llvm_build/./lib/x86_64-unknown-linux-gnu -lc++ -lc++abi -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir/t.tmp.exe
# EXECUTED AS
/b/s/w/ir/cipd_bin_packages/cpython3/bin/python3.8 /b/s/w/ir/x/w/llvm-llvm-project/libcxx/utils/run.py --execdir /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir --  /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir/t.tmp.exe
# executed command: /b/s/w/ir/cipd_bin_packages/cpython3/bin/python3.8 /b/s/w/ir/x/w/llvm-llvm-project/libcxx/utils/run.py --execdir /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir -- /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/test/std/time/time.zone/time.zone.timezone/time.zone.members/Output/get_info.local_time.pass.cpp.dir/t.tmp.exe
# .---command stderr------------
# | Assertion failure: lhs.end == rhs.end /b/s/w/ir/x/w/llvm-llvm-project/libcxx/test/std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp 77
# | 
# | End:
# | Expected output 1880-08-02 00:25:21
# | Actual output   1880-08-02 00:25:00
# `-----------------------------
# error: command failed with exit status: 250
```

Failed build task: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8745592240286102897/overview

This test fails about 1 in 3 runs, randomly after this PR was landed. 
Could you take a look please?

**Comment by vitalybuka** [comment]
Also here https://lab.llvm.org/buildbot/#/builders/236/builds/11695

**Comment by vitalybuka** [comment]
> Also here https://lab.llvm.org/buildbot/#/builders/236/builds/11695
> 
> Test was disabled with [503fb1a](https://github.com/llvm/llvm-project/commit/503fb1aaf4b322173721fbf3855f2168cf6b230b) by @fmayer

@fmayer do you know the reason why it hangs?
If not, maybe correct approach revert this patch and investigate.  Downstream hwasan+libc++ users may hit in own programs.

**Comment by fmayer** [comment]
> @fmayer do you know the reason why it hangs? If not, maybe correct approach revert this patch and investigate. Downstream hwasan+libc++ users may hit in own programs.

I don't, revert SGTM

**Comment by ilovepi** [comment]
@mordante do you think this could be related to https://github.com/llvm/llvm-project/issues/89629? I'm just catching up on this, so I'm not sure yet.

**Comment by vitalybuka** [comment]
@fmayer Would you like to revert to fix the bot? Seems unsupported didn't work anyway.

**Comment by mordante** [comment]
> We got the same test failure on Amrv7 Win-to-Linux cross builder https://lab.llvm.org/buildbot/#/builders/60/builds/17468
> 
> `llvm-libc++-static.cfg.in :: std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp`
> 
> ```
> "C:/Python310/python.exe" "C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/utils/ssh.py" --host=ubuntu@jetson6.lab.llvm.org --execdir C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\time\time.zone\time.zone.timezone\time.zone.members\Output\get_info.local_time.pass.cpp.dir --  C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\time\time.zone\time.zone.timezone\time.zone.members\Output\get_info.local_time.pass.cpp.dir\t.tmp.exe
> # executed command: C:/Python310/python.exe C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/utils/ssh.py --host=ubuntu@jetson6.lab.llvm.org --execdir 'C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\time\time.zone\time.zone.timezone\time.zone.members\Output\get_info.local_time.pass.cpp.dir' -- 'C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\time\time.zone\time.zone.timezone\time.zone.members\Output\get_info.local_time.pass.cpp.dir\t.tmp.exe'
> # .---command stderr------------
> # | Assertion failure: lhs.result == rhs.result C:\buildbot\as-builder-1\x-armv7l\llvm-project\libcxx\test\std\time\time.zone\time.zone.timezone\time.zone.members\get_info.local_time.pass.cpp 91
> # | 
> # | Result:
> # | Expected output 0
> # | Actual output   -1
> # `-----------------------------
> # error: command failed with exit status: 255
> ```
> 
> The target system is Ubuntu 18.04.6 LTS (armv7l).

Which version of the timezone database is installed on this system? I also didn't get a blame e-mail for this build failure? Are the e-mails enabled?

**Comment by mordante** [comment]
> > Also here https://lab.llvm.org/buildbot/#/builders/236/builds/11695
> > Test was disabled with [503fb1a](https://github.com/llvm/llvm-project/commit/503fb1aaf4b322173721fbf3855f2168cf6b230b) by @fmayer
> 
> @fmayer do you know the reason why it hangs? If not, maybe the correct approach is to revert this patch and investigate. hwasan+libc++ users may hit this issue in their programs.

The test is not the fastest, so enabling sanitizer might trigger a time-out. This feature is still experimental in libc++ and I've not looked at performance improvements. (I know there are several improvements that can be done, but I first want working code and benchmarks.)

**Comment by vitalybuka** [comment]
> Did you have time to look at #89629 in more detail?

That's why revert should be the default action here.

**Comment by vitalybuka** [comment]
> The test is not the fastest, so enabling sanitizer might trigger a time-out. This feature is still experimental in libc++ and I've not looked at performance improvements. (I know there are several improvements that can be done, but I first want working code and benchmarks.)

It can be the case, it's flaky with asan as well https://lab.llvm.org/buildbot/#/builders/239/builds/7415
and when it pass it's very slow with asas

```
--------------------------------------------------------------------------
1045.64s: llvm-libc++-shared.cfg.in :: std/time/time.zone/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp
1004.27s: llvm-libc++-shared.cfg.in :: std/atomics/atomics.ref/compare_exchange_weak.pass.cpp
997.76s: llvm-libc++-shared.cfg.in :: std/atomics/atomics.ref/compare_exchange_strong.pass.cpp
303.88s: llvm-libc++-shared.cfg.in :: std/atomics/atomics.ref/wait.pass.cpp
...
```

If we confident that this is just timeout, we need to split or speedup this test. Long tests also a problem, less serious, but still.

**Comment by ilovepi** [comment]
> > @mordante do you think this could be related to #89629? I'm just catching up on this, so I'm not sure yet.
> 
> Did you have time to look at #89629 in more detail?

yes, but I was never able to reproduce it. Our bots do it reliably, but for whatever reason, I wasn't able to, even using docker and following our bot recipe for building clang.

**Comment by ilovepi** [comment]
> > Did you have time to look at #89629 in more detail?
> 
> That's why revert should be the default action here.

agreed

**Comment by vvereschaka** [comment]
@mordante,
>>The target system is Ubuntu 18.04.6 LTS (armv7l).
>Which version of the timezone database is installed on this system? 

`2023c`
```
tzdata/bionic-updates,bionic-security,now 2023c-0ubuntu0.18.04 all [installed]
```
>I also didn't get a blame e-mail for this build failure? Are the >e-mails enabled?
yes, should be enabled.

**Comment by mordante** [comment]
> > > @mordante do you think this could be related to #89629? I'm just catching up on this, so I'm not sure yet.
> > 
> > 
> > Did you have time to look at #89629 in more detail?
> 
> yes, but I was never able to reproduce it. Our bots do it reliably, but for whatever reason, I wasn't able to, even using docker and following our bot recipe for building clang.

Thanks for the information. I also haven't been able to reproduce it.

> @mordante,
> 
> > > The target system is Ubuntu 18.04.6 LTS (armv7l).
> > > Which version of the timezone database is installed on this system?
> 
> `2023c`

Thanks that's quite recent, I will test with this version locally. 

> ```
> tzdata/bionic-updates,bionic-security,now 2023c-0ubuntu0.18.04 all [installed]
> ```
> 
> > I also didn't get a blame e-mail for this build failure? Are the >e-mails enabled?
> 
> yes, should be enabled.

That's odd I can't find a message, maybe it got lost.

For a similar test I have disable testing on `// XFAIL: target={{armv(7|8)l-linux-gnueabihf}}`, does that match this machine's target too?

**Comment by vvereschaka** [comment]
>For a similar test I have disable testing on // XFAIL: target={{armv(7|8)l-linux-gnueabihf}}, does that match this machine's target too?

Do you mean `std\time\time.zone\time.zone.timezone\time.zone.members\sys_info.zdump.pass.cpp` test? No, not really.
The target triple on our armv7 system is `armv7-unknown-linux-gnueabihf`. This regex should covet that triple either: `armv(7|8)(l)?-(.+-)?linux-gnueabihf`.

but anyway, this test gets unsupported for any cross compiled tests there because of `has-no-zdump` feature. A way of detection for this feature does not work correctly for the remote targets. We run the test on the remote boards, but the testsuite is trying to detect `zdump` on the host system.

**Comment by zeroomega** [comment]
I looked into the test flakiness introduced by this PR and filed the bug: https://github.com/llvm/llvm-project/issues/95612 . 

TL;DR is the test results depends on the host's timezone database (which is located at `/usr/share/zoneinfo/tzdata.zi` on Debian). This needs to be addresses as libcxx tests like this should be hermetic. We should either check-in a reference timezone data base in LLVM tree for testing purpose or maybe generate one during the build, instead reading it from the host system.

**Comment by mgorny** [comment]
I'm also seeing it fail, on Gentoo with timezone-data 2024a-r1.

```
# .---command stderr------------
# | Assertion failure: lhs.result == rhs.result /var/tmp/portage/sys-libs/libcxx-19.0.0_pre20240615/work/libcxx/test/std/time/time.zone
/time.zone.timezone/time.zone.members/get_info.local_time.pass.cpp 92
# | 
# | Result:
# | Expected output 0
# | Actual output   -1
# `-----------------------------
# error: command failed with exit status: 250
```

FWICS it's failing on `Indian/Kerguelen`.

**Comment by mordante** [comment]
I really don't see how this test can be flaky. However it seems I underestimated how volatile the time zone database is and maybe we indeed should test with a fixed version. I'll discuss that with @ldionne. For now I created #95659 which marks the test as unsupported. If this passes our CI I'll land it.

**Comment by mgorny** [comment]
If you tell me how I could try giving you more info. However, all the previous experiences prove, relying on timezone-data is extremely flaky, especially that these things change all the time.

**Comment by mordante** [comment]
Thanks for the offer!
Since it is more volatile than I expected I think we should look at using a fixed data set instead of the database provided by the system.
